File: src/0x09강 - BFS/1012_유기농배추.cpp
#include<iostream>
#include<queue>
#define MAX 50
using namespace std;

void bfs(int r, int c, int n, int m, int graph[][MAX])
{
    queue<pair<int,int>> q;
    q.push({r, c});
    graph[r][c] = 0;

    int dr[4] = {-1, 1, 0, 0};
    int dc[4] = {0, 0, -1, 1};

    while(!q.empty())
    {
        int r = q.front().first;
        int c = q.front().second;
        q.pop();

        for(int i=0; i<4; i++)
        {
            int nr = r+dr[i];
            int nc = c+dc[i];

            if(nr>=0&&nr<n&&nc>=0&&nc<m)
            {
                if(graph[nr][nc]==1)
                {
                    q.push({nr,nc});
                    graph[nr][nc]=0;
                }
            }
        }
    }
}
int main()
{
    int t;
    cin >> t;
    for(int i=0; i<t; i++)
    {
        int m, n, k;
        int graph[MAX][MAX] = {0, };
        cin >> m >> n >> k;
        for(int j=0; j<k; j++)
        {
            int x, y;
            cin >> x >> y;
            graph[y][x] = 1;
        }

        int ans=0;
        for(int r=0; r<n; r++)
        {
            for(int c=0; c<m; c++)
            {
                if(graph[r][c]==1)
                {
                    ans++;
                    bfs(r, c, n, m, graph);
                }
            }
        }
        cout << ans << endl;
    }

}

File: src/0x09강 - BFS/1012_유기농배추.py
'''
어떤 배추에 배추흰지렁이가 한 마리라도 살고 있으면 이 지렁이는 인접한 다른 배추로 이동할 수 있어, 그 배추들 역시 해충으로부터 보호받을 수 있다.
'''
from collections import deque


def bfs(x, y, graph):
    q = deque([(x, y)])
    graph[x][y] = 0

    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]

    while q:
        x, y = q.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx < n and 0 <= ny < m:
                if graph[nx][ny] == 1:
                    q.append((nx, ny))
                    graph[nx][ny] = 0


t = int(input())
for _ in range(t):
    m, n, k = map(int, input().split())
    graph = [[0] * m for _ in range(n)]
    for _ in range(k):
        y, x = map(int, input().split())
        graph[x][y] = 1

    ans = 0
    for i in range(n):
        for j in range(m):
            if graph[i][j] == 1:
                ans += 1
                bfs(i, j, graph)

    print(ans)
